### ISSUE TITLE
Extreme HeWD masses

### ISSUE THREAD

**Issue by ilyamandel** [TeamCOMPAS/COMPAS#605]
COMPAS can apparently produce 1.96 solar mass HeWDs (originally pointed out by @solfreludio , I've now confirmed this).  Something is rotten in the state of COMPAS low-mass binary evolution... ;-)

**Comment by nrsegovia** [comment]
For a simple run of 10000 systems there are HeWDs as massive as 2.5 solar mass. My guess is that those stars also continue accretion without changing class, as a HeWD forms when a star on the FGB or HG loses its envelope and has a core mass lower than `MHeF`. This limit is the one Ilya mentioned (Hurley et al. 2000 eq. 2, with the default values in COMPAS). Maybe #512 related? 

**To reproduce**
```
COMPAS --minimum-secondary-mass 0.1 --initial-mass-max 10 --number-of-systems 10000 --random-seed 15
```
**Screenshot**

Taking only systems that create binary WD pairs and checking the mass distribution of the primary/secondary produces the following: 

![Default_1](https://user-images.githubusercontent.com/64709690/128612787-7f0dda90-6192-4093-8d3f-fa6a9596c5fd.png)

**Comment by SimonStevenson** [comment]
Hi @solfreludio is it possible to plot the "detailed output" (--detailed-output) of one of the binaries that produces a massive HeWD? I think you are probably correct that this is due to accretion, but that would be good to see. 

If so, it is not really surprising to see this, since we don't have any prescription for stable burning rates on WDs implemented (as discussed in #512). See also the section "Steady Accretion on to Degenerate Objects" in Hurley et al. 2002 (BSE, https://arxiv.org/pdf/astro-ph/0201220.pdf) for details of a possible prescription for dealing with this. A couple of things jump out: 

1. They have an (arbitrary) accretion efficiency parameter $\epsilon = 0.001$ that limits accretion during novae.
2. When an accreting HeWD reaches 0.7\,M_\odot, they assume that it explodes and leaves no remnant

It would be good to implement both of these as program options that can be varied, with the BSE values as defaults.

**Comment by ashruiter** [comment]
Hi @SimonStevenson, thanks for the feedback! I think we will be able to come up with some prescription(s) to help these poor WDs out 😅 Nicolas @solfreludio is finding some weird things indeed. We'll get there! Really appreciate the quick response!

**Comment by nrsegovia** [comment]
Hi @SimonStevenson, thanks for the insight. I have attached two figures I made, the first one corresponds to the complete evolution of the system over time and the second one is just a plot of the time near the primary evolving into a HeWD. 

For reference, each panel contains:
* (1) : Evolutionary stages.
* (2) : Ratio Radius/Roche_Lobe, also includes information on common envelope and Roche lobe overflow events.
* (3) : MASS_TRANSFER_DIFF, mass accreted/donated per transfer episode.
* (4) : MASS_TRANSFER_TRACKER_HISTORY flags.
* (5) : Mass info.
* (6) : Mass loss rate (MDOT)

It seems that the initial HeWD (primary) mass is *okay* but accretion turns it into such a massive system. Also, am I right to assume that the MT_History is unusual? To me it seems that mass transfer occurs from 2 to 1 instead (FYI the MT History flags are 0,1 and 3). There's the MDOT definition as well... should I understand it as positive means losing mass? In that case it would also be strange as the primary gains mass over time from ~2.32 to ~2.41,M_\odot, before a final *spike* (which I think could come from the envelope of the secondary that is lost when it becomes a WD as well) that gives the final ~ 2.6,M_\odot.
 
![MassiveWD_1](https://user-images.githubusercontent.com/64709690/128903218-f432afcf-5a00-48f2-8b23-52dc20a38aca.png)

![MassiveWD_2](https://user-images.githubusercontent.com/64709690/128903259-1d7f86ce-ab55-4d2d-8864-f1c23441409f.png)

**Comment by ilyamandel** [comment]
@solfreludio , @ashruiter -- very sorry for the long silence on this.  I've been overwhelmed with homeschooling and various other obligations.  At least some of the obligations have been taken care of (alas, not all, and homeschooling is coming back on Monday), but I'm trying to get back to at least a bit of coding, and this issue is a top priority.  

@solfreludio -- suggestion for the future: when you investigate a particular system, it's very helpful if you can either list its random seed and code version, or provide the initial conditions (grid-file format) so others can reproduce the problem.

**Comment by nrsegovia** [comment]
@ilyamandel sorry for that, here are the details for the system in my plots:
```
COMPAS --initial-mass-1 6.89589578359158 --initial-mass-2 2.01573020843856 --number-of-systems 1 --random-seed 3329 --semi-major-axis 1.10640324835946 --eccentricity 0  --detailed-output TRUE --timestep-multiplier 0.01
```
This was done using `COMPAS v02.20.00`, I have not tried running the same system under the newest version.

**Comment by ilyamandel** [comment]
Thank you, @solfreludio !  Something is definitely rotten in the state of <s>Denmark</s> COMPAS... Stay tuned (alas, until after 3 hours of homeschooling...).

Time (Myr),   Event,  M1 (M_o),   type1,  M2 (M_o),   type2,  a (R_o),    e
0     start: Z=0.0142      6.8959      1      2.0157      1      237.9121      0
51.076     star 1 1->2   6.8634      2      2.0157      1      238.7822      0
51.2489     star 1 2->3   6.8628      3      2.0157      1      238.7982      0
51.2722  CE from 1 to 2  1.3234      10      2.0157      1      2.3451      0
51.2722     star 1 3->10   1.3234      10      2.0157      1      2.3451      0
51.2922  Stable MT from 1 TO 2  2.3149      10      1.0243      1      2.9629      0

**Comment by ilyamandel** [comment]
For what it's worth, I think what happens is that there is RLOF immediately after the common envelope [edit: confirmed, that is what happens], leading to stable MT, but that RLOF should be from 2 to 1 in this case -- so need to figure out why it's recorded as 1->2.  That's separate from the issue of dumping this much mass onto a WD...

[Edit (last before homeschooling) -- after the CE, radius_1 / RocheLobe_1 is tiny, as expected, and the MT is definitely cause by radius_2 / RocheLobe_2 exceeding 1, so at least that part is a MT_history mislabeling problem, not a MT implementation problem per se. ]

**Comment by ilyamandel** [comment]
P.S. RLOF immediately after CE keeps on confusing people.  We don't think that's real and usually take it out in post-processing, and I suggest we just assume that such systems lead to merger by default.  I'll make a separate issue for that.

**Comment by ilyamandel** [comment]
Update:  
There are three separate issues. :-)
A. There is a mistake in the MassTransferTrackerHistory.  It doesn't affect the evolution, just a labelling mistake.  I'll upload a PR to fix that shortly.
B. This particular binary entered RLOF right after CE.  We normally discard such binaries in post-processing based on their immediateRLOFPostCEE flag.  To avoid confusion, I propose we change the default behavior to label such binaries as mergers immediately.  I'll create a separate issue to allow people to comment, since it is a change in the defaults.
C. That leaves the question of what to actually do with WDs when they accrete. :)  I can implement the Hurley defaults for now, as Simon suggested above, but it would be great to have @ashruiter 's input on this one!

**Comment by ilyamandel** [comment]
OK, final update on this for today:
A. Fixed, see #636 .
B. I'll wait for a week to give people a chance to comment on this change (see #637 ), then implement it.  For now, I recommend ignoring all systems (like this one) with immediateRLOFPostCEE or optimisticCE flags on -- in our default models, we think they lead to mergers.
C. On reflection, I'd like to do accretion onto WDs properly, rather than implementing a 20-year-old prescription that may be out of date (from Hurley+, 2002, section 2.6.6, steady accretion onto degenerate objects).  Therefore, will now close this issue and open a new one, specifically asking for improved WD accretion prescriptions.
